pkg/checks/external/main.go

```diff
@@ -411,123 +411,94 @@ func (ext *Checker) watchForCheckerPodShutdown(shutdownEventNotifyC chan struct{
 	listOptions := metav1.ListOptions{
 		LabelSelector: kuberhealthyRunIDLabel + "=" + ext.currentCheckUUID,
 	}
+	waitForDeleteChan := make(chan error)
 
-	// start a new watcher with the api
-	watcher := ext.startPodWatcher(listOptions, ctx)
-
-	// use the watcher to wait for a deleted event
-	sawRemovalChan := make(chan struct{}, 2) // indicates that the watch saw the pod be removed
-	stoppedChan := make(chan struct{}, 2)    // indicates that the watch stopped for some reason and needs restarted
-	go ext.waitForDeletedEvent(watcher.ResultChan(), sawRemovalChan, stoppedChan)
-
-	// whenever this func ends, remember to clean up the watcher if its provisioned
-	defer func() {
-		if watcher != nil {
-			watcher.Stop()
-		}
-	}()
+	// start a new watcher with the api and give it a context for aborting early
+	watcher, err := ext.startPodWatcher(listOptions)
+	if err != nil {
+		waitForDeleteChan <- fmt.Errorf("error creating pod watcher: %w", err)
+	}
 
-	// wait to see if the removal happens or the abort happens first by listening for events from the waiter
-	for {
+	// in the background, watch for the deleted event and notify the channel
+	go func() {
+		// watch for either an abort from upstream or removal of the selected pods
 		select {
-		case <-stoppedChan: // the watcher has stopped
-			// re-create a watcher and restart it if it closes for any reason
-			watcher = ext.startPodWatcher(listOptions, ctx)
-			go ext.waitForDeletedEvent(watcher.ResultChan(), sawRemovalChan, stoppedChan) // restart the watch
-		case <-sawRemovalChan: // we saw the watched pod remove
+		case <-ctx.Done(): // graceful shutdown signal
+			ext.log("pod shutdown monitor stopping gracefully")
+		case <-ext.waitForDeletedEvent(watcher): // we saw the watched pod remove
```

Comment:
it's possible that `waitForDeletedEvent(watcher)` can return an error to this channel when a `Watch.Error` event occurs. In that case, should the we recreate the `watcher` and call `waitForDeletedEvent(watcher)` again?

Reply:
I think it's too hard to implement retries here because we won't know if the error is because the supplied watcher is completely closed, or some other condition.  As long as we pass the error upstream, it's probably better to do the retry logic at a higher level in the code.

I have had bad experiences trying to make every thing that tries anything do retries...